### PR TITLE
Refactor Todo settings, offer pre-defined list of Todo apps

### PR DESCRIPTION
**Pull request by yourcontact**
### Description

As per #418, in `Todo Server` setting, I substituted the field where users had to enter the url of the todo app they use with a dropdown menu displaying a list of todo apps:

- [Todoist](https://todoist.com/)
- Franz Todo
- [TickTick](https://www.ticktick.com/)
- Microsoft To Do
- [Habitica](https://habitica.com)
- [Nozbe](https://nozbe.com/)
- [Remember The Milk](https://www.rememberthemilk.com)
- [Any.do](https://www.any.do/ ) (the log in page isn't responsing in Ferdi todo's webview)

Please let me know if you want to delete any of the above, or add any.

### Motivation and Context

The change adds convenience for users.
Now, users can select the todo app they use instead of fetching and copying/pasting the url of the todo app they use.

### How Has This Been Tested?

I tested in Electron. I am not sure how to test it in as a new user cc @kytwb 

### Screenshots (if appropriate):

Kapture 2020-03-24 at 22 38 11

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

That is a refactoring.

### Checklist:

- [ ] My code follows the code style of this project (run `$ yarn lint`). 

@kytwb @vantezzen does my refactoring require a change to the documentation?

### CODE REVIEW

**Comment by kytwb** [comment]
Thanks for picking this one up @yourcontact 💪
Looking forward to the rest and happy to assist again whenever 🙏

**Comment by vantezzen** [comment]
Looks great.

A suggestion would be to add another option "Other service" that displays a text input box below the `select`-element so users can use any service they want.
This would need to be implemented similar to how the darkmode setting displays more sub-settings when you activate it ([Component](https://github.com/getferdi/ferdi/blob/develop/src/components/settings/settings/EditSettingsForm.js#L462-L476), [Container](https://github.com/getferdi/ferdi/blob/develop/src/containers/settings/EditSettingsScreen.js#L548)).
This would also require a change to [the webview](https://github.com/getferdi/ferdi/blob/develop/src/features/todos/components/TodosWebview.js#L226) to use that custom value istead if selected.

**Comment by yourcontact** [comment]
Thank you @vantezzen.

I would like to share an update on where I am with the refactoring Ferdi Todo setting:

- added "Other service" to the drop down menu
- added a field that appears when "Otherservice" is selected in the dropdown menu

Next step is to fix the field for users to input URL of the todo service/app they use. The field currently collapse when I try to input data.

HptD0U

File: src/components/settings/settings/EditSettingsForm.js [hunk 1]
```diff
@@ -34,7 +34,7 @@ const messages = defineMessages({
   },
   headlineGeneral: {
     id: 'settings.app.headlineGeneral',
-    defaultMessage: '!!!General',
+    defaultMessage: '!!!Generadddl',
```
Comment (vantezzen) [inline src/components/settings/settings/EditSettingsForm.js]: Revert this change

File: src/components/settings/settings/EditSettingsForm.js [hunk 2]
```diff
@@ -333,28 +333,37 @@ export default @observer class EditSettingsForm extends Component {
...
-                  style={{
-                    borderTop: 0, marginTop: 0, paddingTop: 0, marginBottom: '2rem',
-                  }}
-                >
-                  { intl.formatMessage(messages.todoServerInfo) }
-                </p>
+                {form.$('enableTodos').value && (
+                  <div>
+                    <Select field={form.$('todoServer')} />
+                    {form.$('todoServer').value === 'otherService' && (
```
Comment (vantezzen) [inline src/components/settings/settings/EditSettingsForm.js]: You should do this in the container element (src/containers/settings/EditSettingsScreen) and pass something like `idTodosActivated` and `isUsingCustomTodoService` prop to the component instead - at least this is how we've done it for most other settings like this.

File: src/i18n/locales/af.json [hunk 3]
```diff
@@ -272,7 +272,7 @@
   "settings.app.form.accentColor": "Accent color",
   "settings.app.form.adaptableDarkMode": "Synchronize dark mode with my Mac's dark mode setting",
   "settings.app.form.autoLaunchInBackground": "Open in background",
-  "settings.app.form.autoLaunchOnStart": "Launch Ferdi on start",
+  "settings.app.form.autoLaunchOnStart": "Launch sdfsdf Ferdi on start",
```
Comment (vantezzen) [inline src/i18n/locales/af.json]: Revert that change

File: src/i18n/locales/en-US.json [hunk 4]
```diff
@@ -302,7 +302,7 @@
   "settings.app.form.showDisabledServices": "Display disabled services tabs",
   "settings.app.form.showMessagesBadgesWhenMuted": "Show unread message badge when notifications are disabled",
   "settings.app.form.startMinimized": "Start minimized in tray",
-  "settings.app.form.todoServer": "Todo Server",
+  "settings.app.form.todoServer": "Todo Service",
```
Comment (vantezzen) [inline src/i18n/locales/en-US.json]: You should also change this in https://github.com/getferdi/ferdi/blob/cecdbc4689/src/containers/settings/EditSettingsScreen.js#L85

File: src/components/settings/settings/EditSettingsForm.js (hunk 1 above)
Comment (yourcontact) [reply src/components/settings/settings/EditSettingsForm.js]: I fixed that typo https://github.com/yourcontact/ferdi/commit/62f1a167bb56a5f62a56bcd88fd67f6341e9dc32

File: src/i18n/locales/en-US.json (hunk 4 above)
Comment (yourcontact) [reply src/i18n/locales/en-US.json]: I reverted that change

**Comment by yourcontact** [comment]
@vantezzen I resolved your comments https://github.com/getferdi/ferdi/pull/477#pullrequestreview-378435525 and updated my initial PR comment https://github.com/getferdi/ferdi/pull/477#issue-390407136 with latest change. 

I am not sure how to fix this (cc @kytwb):
Screen Shot 2020-03-24 at 10 48 35 PM

**Comment by kytwb** [comment]
@yourcontact can you try removing these files?

Screen Shot 2020-03-25 at 12 43 28 AM

**Comment by kytwb** [comment]
@yourcontact feel free to post updates in new comments instead of editing past comments, it's less confusing for the thread history 🙏 no big deal though

**Comment by vantezzen** [comment]
We will still need a migration from the old settings to the new ones in https://github.com/getferdi/ferdi/blob/develop/src/stores/SettingsStore.js#L185. I had a custom Todo server set but the settings screen says that I selected Todoist:

Screenshot 2020-03-25 at 10 42 49

**Comment by kytwb** [comment]
@vantezzen Good catch! 
Regarding the errors, we should probably try to load the Custom Todo URL only once it's a valid URL.

**Comment by yourcontact** [comment]
> @yourcontact can you try removing these files?
> 
> Screen Shot 2020-03-25 at 12 43 28 AM

I removed those files.

**Comment by yourcontact** [comment]
@kytwb and I added a regex to validate the URL when a user select `Other service`.

>We will still need a migration from the old settings to the new ones in https://github.com/getferdi/ferdi/blob/develop/src/stores/SettingsStore.js#L185.

@kytwb will take over this pull request and set a migration from old settings.

**Comment by yourcontact** [comment]
@vantezzen I will work on the migration of old settings with @kytwb tomorrow.

**Comment by kytwb** [comment]
I had to get going with my day, so we only worked on https://github.com/getferdi/ferdi/pull/506; we're pairing again tomorrow at 9am (my local time) to address the missing migration.

**Comment by yourcontact** [comment]
I resolved conflicts in src/containers/settings/EditSettingsScreen.js.

For the conflicts in the two following files:
- src/i18n/locales/defaultMessages.json;
- src/i18n/messages/src/containers/settings/EditSettingsScreen.json;

It looks like there are conflicts due to code style changes. 

Shall I apply what is in the "refactor/settings-todo" branch and delete what is on the "develop" branch?

**Comment by kytwb** [comment]
I am not sure myself what's the best thing to do right now. Can you try keeping your changes and re-running `npm run manage-translations`?

**Comment by yourcontact** [comment]
> I am not sure myself what's the best thing to do right now. Can you try keeping your changes and re-running npm run manage-translations?

@kytwb I ran `npm run manage-translations` on the "refactor/settings-todo" branch.
Below is a screenshot of what I got. I am not sure what to do next.
Screen Shot 2020-03-31 at 12 08 06 PM

**Comment by kytwb** [comment]
Check if the diff looks good to you (`git diff`), then you can finalize the merge with `git commit --no-edit`.

**Comment by kytwb** [comment]
Not sure what's the best way to handle this given the restrictions enforced (scenario where we need to help in pull requests from external contributors), but I think it's best for the long term if we figure it out.

**Comment by kytwb** [comment]
Closed by #527.
